larpandoradlcontent/LArTwoDReco/DLThreeDClusterSplittingAlgorithm.cc

```diff
@@ -0,0 +1,1350 @@
+        this->GetPredictedSplitIndices(features, splitIndices);
+        
+        // Remove any that are too close to the nu vertex (network picks up on hit sharing)
+        const HitType hitType(LArClusterHelper::GetClusterHitType(pCluster));
+        this->FilterSplitIndices(clusterPath, hitType, features, splitIndices);
+        
+        if (splitIndices.empty())
+            return;
+
+        // Get split positions
+        for (const int splitIndex : splitIndices)
+        {
+            CartesianVector position(clusterPath.at(splitIndex).m_pHit->GetPositionVector());
+            splitPositions.push_back(position);
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DLThreeDClusterSplittingAlgorithm::FindPath(const Cluster *const pCluster, const TwoDSlidingFitResult &clusterFit, ClusterPath &clusterPath) const
+{
+    // Get cluster hits
+    CaloHitList clusterHits;
+    LArClusterHelper::GetAllHits(pCluster, clusterHits);
+
+    // Fit l decomposition
+    for (const CaloHit *const pCaloHit : clusterHits)
+    {
+        float thisHitL(0.f), thisHitT(0.f);
+        clusterFit.GetLocalPosition(pCaloHit->GetPositionVector(), thisHitL, thisHitT);
+        const int lBinIndex(std::floor(thisHitL / m_lBinSize));
+
+        ClusterHit clusterHit(pCaloHit, lBinIndex, thisHitT);
+
+        const auto iter(std::find(clusterPath.begin(), clusterPath.end(), clusterHit));
+
+        if (iter == clusterPath.end())
+        {
+            clusterPath.emplace_back(clusterHit);
+        }
+        else if (thisHitT < iter->m_t)
+        {
+            *iter = clusterHit;
+        }
+    }
+
+    // Order low->high l
+    std::sort(clusterPath.begin(), clusterPath.end(), [](const ClusterHit &lhs, const ClusterHit &rhs) { return lhs.m_l < rhs.m_l; });
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DLThreeDClusterSplittingAlgorithm::InitialiseFeatures(Features &features) const
+{
+    features.insert(std::make_pair("Longitudinal", Feature((0.0), 1.0, std::vector<float>())));
+    features.insert(std::make_pair("Transverse", Feature((-0.01), 2.57, std::vector<float>())));
+    features.insert(std::make_pair("Energy", Feature(0.49, 0.28, std::vector<float>())));
+    features.insert(std::make_pair("HitWidth", Feature(0.56, 0.16, std::vector<float>())));
+    features.insert(std::make_pair("Theta", Feature((-0.01), 0.11, std::vector<float>())));
+    features.insert(std::make_pair("SecVertex", Feature(93.19, 103.39, std::vector<float>())));
+    features.insert(std::make_pair("GapSep", Feature(0.0, 1.0, std::vector<float>())));
+    features.insert(std::make_pair("EventHitSep", Feature(5.00, 5.83, std::vector<float>())));
+    features.insert(std::make_pair("ClusterHitSep", Feature(0.54, 0.13, std::vector<float>())));
```

Comment:
Magic numbers. Where do these come from, and in any case, it feels like these should probably be configurables?